### ISSUE TITLE
Supporting Qt 6

### ISSUE THREAD

**Issue by The-Compiler** [qutebrowser/qutebrowser#5395]
~~The next Qt release after 5.15 will be Qt 6. The [initial schedule](https://lists.qt-project.org/pipermail/development/2020-April/039382.html) plans for a release 2020-12-01, with feature freeze on 2020-08-31.~~

QtWebEngine is coming back with Qt 6.2. [Release schedule](https://wiki.qt.io/Qt_6.2_Release):

- Feature freeze 2021-06-07
- Alpha 2021-06-23
- Beta 1 2021-07-6
- Beta 2 2021-07-27
- Beta 3 2021-08-17
- Beta 4 2021-08-31
- RC 2021-09-14
- Final 2021-09-30

Links:

- [Interesting JIRA issues](https://bugreports.qt.io/issues/?jql=component%20%3D%20webengine%20and%20fixVersion%20in%20(6.0%2C%206.0.0%2C%20%226.0.0%20Alpha%22%2C%20%226.0.0%20Beta1%22%2C%20%226.0.0%20Beta2%22%2C%20%226.0.0%20Beta3%22%2C%20%226.0.0%20Beta4%22%2C%20%226.0.0%20Beta5%22%2C%20%226.0.0%20RC%22%2C%20%226.0.0%20RC2%22%2C%206.0.1%2C%206.0.2%2C%206.0.3%2C%206.0.4%2C%206.1%2C%206.1.0%2C%20%226.1.0%20Alpha%22%2C%20%226.1.0%20Beta%201%22%2C%20%226.1.0%20Beta2%22%2C%20%226.1.0%20Beta3%22%2C%20%226.1.0%20FF%22%2C%20%226.1.0%20RC%22%2C%20%226.1.0%20RC2%22%2C%206.1.1%2C%206.1.2%2C%206.1.3%2C%206.2%2C%206.2.0%2C%20%226.2.0%20Alpha%22%2C%20%226.2.0%20Beta1%22%2C%20%226.2.0%20Beta2%22%2C%20%226.2.0%20Beta3%22%2C%20%226.2.0%20Beta3%22%2C%20%226.2.0%20Beta4%22%2C%20%226.2.0%20FF%22%2C%20%226.2.0%20RC%22))
- Interesting codereview changes: [dev](https://codereview.qt-project.org/#/q/project:qt/qtwebengine+branch:dev,n,z), [6.2](https://codereview.qt-project.org/q/project:qt/qtwebengine+branch:6.2)

# API changes

Probably going to be a bunch, but rather small changes... The plan for qutebrowser is to stay compatible with Qt 5 for a while, while gaining compatibility with Qt 6.

- [Rename QWebEngineDownloadItem -> QWebEngineDownloadRequest](https://codereview.qt-project.org/c/qt/qtwebengine/+/309916)
- [Remove obsolete, deprecated api from qwebenginedownloaditem](https://codereview.qt-project.org/c/qt/qtwebengine/+/309741)
- [Unify certificate error handling](https://codereview.qt-project.org/c/qt/qtwebengine/+/309513)
- [Move QMenu handling from page to view](https://codereview.qt-project.org/c/qt/qtwebengine/+/305879)
- [Merge all webenginecontextmenudata wrappers into one class](https://codereview.qt-project.org/c/qt/qtwebengine/+/308959)
- [Improve QWebEngineCertificateError api](https://codereview.qt-project.org/c/qt/qtwebengine/+/309921)
- [Remove default qwebenginesettings getter](https://codereview.qt-project.org/c/qt/qtwebengine/+/310611)
- [Move qwebenginesettings to core](https://codereview.qt-project.org/c/qt/qtwebengine/+/310612)
- [Move QWebEngineFullScreenRequest to core](https://codereview.qt-project.org/c/qt/qtwebengine/+/302366)
- [Move to qwebenginescript to core](https://codereview.qt-project.org/c/qt/qtwebengine/+/311285)
- [Move qtwebenginescriptcollection class to core](https://codereview.qt-project.org/c/qt/qtwebengine/+/311691)
- [Cleanup qwebenginescript and qwebenginescriptcollection api](https://codereview.qt-project.org/c/qt/qtwebengine/+/310838)
- [Drop dependency on default notification presenter in profile](https://codereview.qt-project.org/c/qt/qtwebengine/+/311938)
- [Drop dependency on page in profile](https://codereview.qt-project.org/c/qt/qtwebengine/+/311939) (internal only?)
- [Move profile from widgets to core](https://codereview.qt-project.org/c/qt/qtwebengine/+/311817)
- [Remove deprecated url interoceptor setter from profile](https://codereview.qt-project.org/c/qt/qtwebengine/+/311818)
- [Move client certificate selection to core](https://codereview.qt-project.org/c/qt/qtwebengine/+/316847)
- [Remove qwebengineview setter from qwebenginepage](https://codereview.qt-project.org/c/qt/qtwebengine/+/317989)
- [Move QWebEngineHistory to core](https://codereview.qt-project.org/c/qt/qtwebengine/+/318448)
- [Move QWebEnginePage to core](https://codereview.qt-project.org/c/qt/qtwebengine/+/318518)
- [Add Qt WebEngine version getter](https://codereview.qt-project.org/c/qt/qtwebengine/+/333557)
- [Make handling of certificate errors possible from a signal](https://codereview.qt-project.org/c/qt/qtwebengine/+/346682)
- [Add missing FileSelectionMode enum value](https://codereview.qt-project.org/c/qt/qtwebengine/+/349778)
- [Move createWindow to a signal](https://codereview.qt-project.org/c/qt/qtwebengine/+/346723) (internal only?)
- [Move QQuickWebEngineNavigationRequest to Core](https://codereview.qt-project.org/c/qt/qtwebengine/+/351593)
- [Merge WebEngineHistory's quick implementation into QWebEngineHistory](https://codereview.qt-project.org/c/qt/qtwebengine/+/327625)
- [QWebEnginePage: add new 'loading' property and expose QWebEngineLoadingInfo](https://codereview.qt-project.org/c/qt/qtwebengine/+/352611)
- [Move printing finished callback to signal](https://codereview.qt-project.org/c/qt/qtwebengine/+/353136)
- [Add API for favicon database](https://codereview.qt-project.org/c/qt/qtwebengine/+/309633)
- [Remove left-over deprecated API QWebEngineProfile::setRequestInterceptor()](https://codereview.qt-project.org/c/qt/qtwebengine/+/355922)
- [Add QPageRanges to PDF printing](https://codereview.qt-project.org/c/qt/qtwebengine/+/356189)
- [Make default profile off the record](https://codereview.qt-project.org/c/qt/qtwebengine/+/352441)
- [Remove deprecated useforglobalcertificateverification](https://codereview.qt-project.org/c/qt/qtwebengine/+/358405)

# API review follow ups

- [Remove duplicated 'downloadProgress' in favor of separate signals (Ib825d63c) · Gerrit Code Review](https://codereview.qt-project.org/c/qt/qtwebengine/+/359551)
- [Update parameters to findText callback (I622ff55c) · Gerrit Code Review](https://codereview.qt-project.org/c/qt/qtwebengine/+/359136)
- [Add QWebEngineNewWindowRequest::openIn(QWebEnginePage *) (I15e72d52) · Gerrit Code Review](https://codereview.qt-project.org/c/qt/qtwebengine/+/359814)
- [Add QWebEnginePage::view() replacement (I552f0047) · Gerrit Code Review](https://codereview.qt-project.org/c/qt/qtwebengine/+/361654)
- [Return printToPdf to QWebEnginePage (I0e14563c) · Gerrit Code Review](https://codereview.qt-project.org/c/qt/qtwebengine/+/361156)

# Interesting non-API changes

- [Fix inconsistent number of load signals and their order](https://codereview.qt-project.org/c/qt/qtwebengine/+/340965) (not an API change, but relevant for us)
- [Implement dialogs of File System Access API](https://codereview.qt-project.org/c/qt/qtwebengine/+/353879) (no API change, just potentially interesting)
- [Add basic USB device logging](https://codereview.qt-project.org/c/qt/qtwebengine/+/334655) (same)
- [Suppress error pages also for http errors if they are disabled](https://codereview.qt-project.org/c/qt/qtwebengine/+/324923/) (no API change, but interesting)
- [Disable kAllowContentInitiatedDataUrlNavigations (I9398f54b) · Gerrit Code Review](https://codereview.qt-project.org/c/qt/qtwebengine/+/358628)
- [Widgets (like the command text box) now insert test into the primary selection (if available)](https://bugreports.qt.io/browse/QTBUG-59879) (ignore the fix version, it only made it into 6.0 but didn't get a new ticket)

# Planned changes in PyQt

- `PyQt5` -> `PyQt6` rename (see #995)
- Enums won't be accessible unscoped, [only scoped](https://www.riverbankcomputing.com/static/Docs/PyQt5/gotchas.html#enums) - scoped access is possible since Qt 5.11 I believe, so if we want to access enum members without some unwieldy `enum_value(QWebEnginePage, 'FindFlag', 'FindBackward')` wrapper, we'll need to drop older versions at that point (#3839) - split off to #5904.

More from the website:

> PyQt6 introduces the following incompatibilities over PyQt5:
> 
> - The `Qt` module has been removed.
> - `QtWidgets.qApp` has been removed.
> - The `QtCore.PYQT_CONFIGURATION` dict has been removed.
> - `pyrcc` has been removed.
> - Scoped enums are now implemented as `enum.Enum` (PyQt5 used `enum.IntEnum`). Named enums are now implemented as `enum.IntEnum` (PyQt5 used a custom type). PyQt5 allowed an int whenever an enum was expected. PyQt6 requires the correct type (ie. it will not accept an int instead of an enum).
> - All `exec_()` methods have been removed.
> - All `print_()` methods have been removed.
> - Python v3.6 or later is required.

# Changes which probably didn't make it 

- [Add QWebEngineurlRequestBody (Ia0e4b5a5) · Gerrit Code Review](https://codereview.qt-project.org/c/qt/qtwebengine/+/314533)
- [Add a new signal "launchExternalProtocol" when launching external url (I744b6ca9) · Gerrit Code Review](https://codereview.qt-project.org/c/qt/qtwebengine/+/314607)
- [Clear the cache in time range (If80bba20) · Gerrit Code Review](https://codereview.qt-project.org/c/qt/qtwebengine/+/313190)
- [Add QWebEngineurlRequestBody (Ia0e4b5a5) · Gerrit Code Review](https://codereview.qt-project.org/c/qt/qtwebengine/+/314533)
- [Add QWebEnginePage::httpStatusCode() (I350c003e) · Gerrit Code Review](https://codereview.qt-project.org/c/qt/qtwebengine/+/318313)

**Comment by The-Compiler** [comment]
Apparently the Qt 6 release is going to be "essential" Qt modules only, i.e. not QtWebEngine. That will then be part of Qt 6.1. Instead, the underlying Chromium will be upgraded for Qt 5.15 patch releases, though I'm not sure if those will be commercial-only at the point Qt 6.0 is released...

If Qt 6.1 is following the usual schedule, that'd be around May/June 2021. Probably doesn't make much sense for qutebrowser to start migrating things before that. On the positive side, that buys us some time for all the other things that should be done for a v2.0.0 release.

**Comment by The-Compiler** [comment]
FWIW the initial QtWebEngine API review is out now: https://codereview.qt-project.org/c/qt/qtwebengine/+/312867

**Comment by The-Compiler** [comment]
A [clarification](https://bugreports.qt.io/browse/QTBUG-63235?focusedCommentId=533371&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-533371) on the timeline for QtWebEngine:

> Note QtWebengine also meant to be independent of Qt releases in the same time frame, so might not make that much sense to promise it with Qt releases. We hope to have a QWE release ready around the same time as 6.1, but the official goal is with 6.2.

Judging from the 5.14 release, that'd mean between May/June 2021 (6.1) and December 2021/January 2022 (6.2). Still plenty of time until then... Still no details (as far as I can tell) on what the independent release means for e.g. PyQt (and thus qutebrowser) exactly.

**Comment by The-Compiler** [comment]
From a [Qt Development mailinglist mail](https://lists.qt-project.org/pipermail/development/2020-November/040704.html) today:

> We have long discussed, that the timing of our feature releases to be just before summer and Christmas vacation is a bit unfortunate, as we have little slack for delays without going into the vacation period. Especially the releases in December have sometimes been difficult in that respect. So we’d like to push the schedule a bit and move the minor releases towards a Spring/Autumn schedule.
> 
> A somewhat shorter release cycle directly after 6.0 is probably a good idea anyway, as we will probably still need to do changes/fixes that don’t quite fit with our policy for patch level releases.
> 
> So the idea is to shorten the release cycle for Qt 6.1 a bit and focus mainly on bug fixing and stability for that release. We’d aim for a feature freeze by the end of January, and a final Qt 6.1.0 release end of April.
> 
> 6.2 would then also happen a bit earlier, with a feature freeze in June and a release end of September.
> 
> Content wise, I believe we’ll start seeing more and more of the add-ons from Qt 5 being supported over the next 6-9 months, and I believe that with Qt 6.2 we will have brought most modules that we supported in Qt 5.15 over to Qt 6. 
>

**Comment by The-Compiler** [comment]
There's finally some more official communication around this now: https://www.qt.io/blog/add-on-support-in-qt-6.0-and-beyond

**Comment by The-Compiler** [comment]
Nothing left here for v2.0.0 - I've done some trivial changes; the enums change (#5904) needs bigger changes to the mypy stubs first and won't be in v2.0.0.

**Comment by The-Compiler** [comment]
QtWebEngine API review:

Notable changes I can see:

- New `qWebEngineVersion()` and `qWebEngineChromiumVersion()`
- Some `QWebEngineCertificateError` changes, but IIRC we don't use that yet
- `QWebEngineDownloadItem` renamed to `QWebEngineDownloadRequest`
   * `DownloadType` / `type()` removed, but we don't use that
   * Deprecated `path()` and `setPath()` removed
   * More signals we could potentially use?
   * `finished` signal renamed to `isFinishedChanged`
- `QWebEngineFullScreenRequest` now uses some kind of callback?
- `QWebEngineHistory` changed, now has a `QWebEngineHistoryModel`?
- QWebEnginePage changes
   * `QWebEnginePage::FileSelectionMode::FileSelectUploadFolder`
   * `QWebEnginePage::view()` and `setView()` removed
   * Most callbacks in `QWebEnginePage` changed to `std::function`, but that hopefully doesn't affect us via Python
   * Print API was moved to page?
   * New `certificateError` signal, emitting `QWebEngineCertificateError` (replaces `certificateError` virtual)
   * New `navigationRequested` signal, emitting `QWebEngineNavigationRequest` (new class)
   * New `newWindowRequested` signal, emitting `QWebEngineNewWindowRequest` (new class)
   * New `loadingChanged` signal, emitting `QWebEngineLoadingInfo` (new class)
- `QWebEngineProfile` changes
   * Deprecated `QWebEngineProfile::setRequestInterceptor` removed (qutebrowser should already be adjusted)
   * New `notificationPresenter()` getter
- `QWebEngineScript` changes
   * `isNull()` removed
   * New `sourceUrl()` and `setSourceUrl()`
   * `QWebEngineScriptCollections.size()` removed (duplicates `.count()`)
   * `QWebEngineScriptCollections.findScript()` and `.findScripts()` replaced by `.find()`
- `QWebEngineSettings` changes
   * `globalSettings()` moved (?)
   * `defaultSettings()` removed
   * New `setParentSettings`

**Comment by The-Compiler** [comment]
There's some more progress in the `qt6-test` branch thanks to @toofar - known issues:

- [ ] Various fun around enum handling
- [ ] Missing qutebrowser icon resources (#6062)
- [ ] *The issue you reported about UrlFormattingOption and ComponentFormattingOption isn't entirely resolved, you can combine them now (with `|`) but you can't pass ComponentFormattingOption values to the methods that claim to accept a UrlFormattingOption but actually should accept both.*
- [ ] Category headings in the completion dialog don't look fancy anymore, selected completion row's background is blue
- [ ] `qenum_key` and friends probably can be simplified since we're now dealing with real Python enums.
- [ ] *QWebEngineHistory serialization doesn't work anymore, for instance when closing a tab. A non fatal error from Qt is printed out `QVariant::save: unable to save type 'QObject*' (type id: 39)`. When you try to undo the tab close then the deserialization will choke on the incomplete data stream. The individual history items can be serialized fine though. Saving and loading sessions still seems to work fine. I'm not sure whether this is a Qt or PyQt issue, I thought it might be https://codereview.qt-project.org/c/qt/qtwebengine/+/327625 (parent of QWebEngineHistory is now a QObject) but I tried to reproduce in C++ and didn't get the same error (it also didn't print out as much hex as I was expecting though, so I might have messed something else up).*
- [ ] Lots more in my original changes, I'm sure

**Comment by The-Compiler** [comment]
# New features in Qt 6+

- [[QTBUG-93848] QNetworkInformation: Behind connectivity portal - Qt Bug Tracker](https://bugreports.qt.io/browse/QTBUG-93848)

**Comment by The-Compiler** [comment]
Some rough instructions for trying the `qt6-test` branch if someone feels adventurous:

- Clone the git repo, e.g. via `git clone https://github.com/qutebrowser/qutebrowser.git`
- `cd qutebrowser`
- Check out the branch using `git checkout qt6-test`
- Create a new separate environment via `python3 scripts/mkvenv.py --venv-dir .venv-qt6`
- Run `.venv-qt6/bin/qutebrowser --temp-basedir` and verify via `:version` that you're indeed on Qt 6

Lots of stuff is known to be broken - if you notice something not mentioned in the comment above, feel free to mention it here.

**Comment by The-Compiler** [comment]
I see someone™ has [reported](https://www.riverbankcomputing.com/pipermail/pyqt/2021-November/044393.html) the serialization issue to PyQt and it was [resolved](https://www.riverbankcomputing.com/pipermail/pyqt/2021-November/044399.html) - relevant changelog entry:

```
2021-11-22  Phil Thompson  <phil@riverbankcomputing.com>

	* NEWS, PyQt6.msp:
	Added /Constrained/ to the QVariant serialiser so that it doesn't
	act as a catchall and bypass all subsequent serialisers (including
	slot extenders).
	[ce7d2be67940] <6.2-maint>
```

Makes me wonder if there is some kind of workaround we could do, but guess not...

I've also worked a bit more on the `qt6-test` branch today and noticed some more issues:

- [ ] PyInstaller builds [seem to fail](https://github.com/qutebrowser/qutebrowser/runs/4336651577?check_suite_focus=true) with `RuntimeError: PyQt6.QtWebEngineWidgets cannot import type '\ufffda\ufffd:\ufffd' from PyQt6.QtCore`. No clue what that means...
- [ ] qutebrowser's JS seems to not run or something? e.g. using `f` says "Unknown error while getting elements" and the JS console reveals `Cannot read property 'webelem' of undefined`, which sounds like `window._qutebrowser` isn't defined. Possibly a bug or some change in `QWebEngineScript`? @toofar, did you look into that one so far?
- [ ] We use `.path()` for downloads which is now removed ("`# FIXME use 5.14 API`")

**Comment by toofar** [comment]
> qutebrowser's JS seems to not run

I hadn't noticed. Seems it was just the fault of a hasty hack putting some stuff in different worlds though. Fixed.

I also noticed that

- [ ]: the developer tools don't remember your preferences.

Possibly due to it referring to the default (now off-the-record) profile. I pushed b961c3b8201d987 to make the default profile (our one, used for non-private windows) persistent again. I tried to make the inspector persist it's settings by setting a page linked to _our_ default profile on the WebEngineInspectorView but it didn't help.

**Comment by The-Compiler** [comment]
I pushed some more fixes here and there. I had to change the world ID handling a bit in bf5a49df786c179f71c1fbaf43c074f9cf9a6de9 because the `JsWorld` value already was an int. Fingers crossed that it fixed more than it broke.

Various issues in the tests still due to the QUrl bug. Reported now at [Qt6: QUrl and UrlFormattingOption vs. ComponentFormattingOption](https://www.riverbankcomputing.com/pipermail/pyqt/2021-November/044409.html).

- [ ] Also it looks like we get a segfault on PyQt exceptions, both in the end2end subprocess as well as in unit tests with pytest-qt...

**Comment by The-Compiler** [comment]
> And [ec200c3](https://github.com/qutebrowser/qutebrowser/commit/ec200c3486f6e57c084c947a2661461d709348ae) to let the developer tools use the same profile as the page it is inspecting.

Hmmm, wouldn't it make more sense to always use the default profile for the inspector? The fact that it uses a profile to store its settings is an implementation detail to the user, and I would't expect e.g. the inspector to lose its settings when inspecting a private website. OTOH I guess it depends on whether the stored data contains data related to the inspected website...

**Comment by toofar** [comment]
Yeah I agree it might be a bit surprising, but also easy to explain (it's a third party tool, it's isolated in private mode). Maybe I'm just used it it using the same profile across both types of windows because that is what we have done before. But apart from devtoolspref.json I'm not sure what other data it might store. The only identifying thing I see in there is `resourcesLastSelectedElementPath` which can have a domain in it.
Mostly the "it's more private" justification is a bit retroactive. I didn't really do it that way on purpose. Once I was force to explicitly choose a profile it seemed weird to stick to the persistent one.

**Comment by toofar** [comment]
- [ ]: hopefully I'm not just seeing double but it looks like [QSqlQuery.boundValues](https://doc.qt.io/qt-6/qsqlquery.html#boundValues) returns each bound value twice? For example have it print the bound values [here](https://github.com/qutebrowser/qutebrowser/blob/f1f1eec3fa/qutebrowser/misc/sql.py#L326) and then search for something in the history completion. This has the effect of making [this conditional](https://github.com/qutebrowser/qutebrowser/blob/f1f1eec3faf8a331f2c8a099894006c32c3dd40d/qutebrowser/completion/models/histcategory.py#L107) set a new query in all cases except when you have two search terms. I'm on 6.2.1
- [ ]: I get a segfault when opening a google meet invite. On the pre-meeting "enter your name and configure stuff" page. Jitsi works fine. This was when not logged into any google account (didn't try it logged in). Not sure how to help other people reproduce since I can't share the link and can't generate one. I'll try to make a debug build at some point.

**Comment by The-Compiler** [comment]
I can't reproduce the `boundValues` issue - here's what I get:

```
{':0': '%%'}
{':0': '%h%'}
{':0': '%ht%'}
{':0': '%htt%'}
{':0': '%http%'}
{':0': '%https%'}
{':0': '%httpsf%'}
```

**Comment by The-Compiler** [comment]
Wait, no, I was accidentally testing with Qt 5 there - indeed with Qt 6:

```
['%%', '%%']
['%h%', '%h%']
['%ht%', '%ht%']
['%htt%', '%htt%']
['%http%', '%http%']
['%httpp%', '%httpp%']
```

**Comment by The-Compiler** [comment]
I have a Qt 6 debug build now and could reproduce the Google Meet issue. I reported it upstream: [[QTBUG-99119] [REG 5 -> 6] Segfault in QtWebEngineCore::WebContentsDelegateQt::webEngineSettings() on Google Meet - Qt Bug Tracker](https://bugreports.qt.io/browse/QTBUG-99119)

**Comment by tinywrkb** [comment]
Initial work to [update the Flatpak QtWebEngine base app to 6.2.x](https://github.com/qutebrowser/qutebrowser/issues/5395), and a [PyQt6 branch of the Flatpak qutebrowser app](https://github.com/tinywrkb/org.qutebrowser.qutebrowser/tree/pyqt6).

I haven't had any success yet in building PyQt6 against the KDE 6.2 Flatpak runtime. I'm not sure what's going on, I'm seeing `::QOpenGLFunctions_ES2’ has not been declared` error while `qopenglfunctions_es2.h` is included.  
I tried setting `QT_CONFIG=+opengles2` and `CONFIG+=opengles2` with `--qmake-setting`, but these didn't help.

**Comment by tinywrkb** [comment]
![swappy-20211228-220808](https://user-images.githubusercontent.com/44030034/147603332-d292242f-4f1b-4fe0-925c-7e724c99d261.png)

**Comment by toofar** [comment]
I pushed another couple of fixes, both of which could be cherry picked to master if someone was feeling bold.

* 54b817898f21 adapt to the QSqlQuery.boundValues() API changing (fixes the history completion not updating when you have two search terms). There is no equivalent API so had to achieve it a different way which is backwards compatible. (cc @rcorre in case you think it is easier to just reduce our dependency on boundValues())
* 2866cae6b342 allow sites (like youtube) to request fullscreen again, this broke with Qt6 but only because a long deprecated way of getting screen geometry was dropped

Some other issues I've noticed:

- [x] widgets (like the command mode text entry) now fill X11 primary selection when you highlight text. Actually a feature, not a bug, I'm going to have to adapt some of my URL editing habits though. I added it to the "Interesting non-API changes" in the issue description
- [ ] download bar at the top of the screen pushes the page viewport down (instead of resizing it) so the bottom of it isn't inside the browser anymore (TODO: confirm once I have a proper QNetwork stack)
- [ ] favicons don't show for some tabs sometime (TODO: find reproducer, going to eg https://github.com/qutebrowser/qutebrowser/commits/qt6-test and clicking on one of the commits seems to lose the favicon)
- [ ] launching a userscript throws an exception relating to QUrl formatting flags, probably fixed already and I just need to update my PyQt

I'll (at some point) look at building all the latest stuff and see if the developer tools element inspector works better now and if I can get my downloads work.

**Comment by The-Compiler** [comment]
Some more I've noticed:

- [x] Clicking a download crashed, fixed in 6344e309c77079b443d9b087edb5969cb5a7d1bb.
- [ ] On dark pages like https://start.duckduckgo.com/?kae=d I sometimes see white squares popping up
- [ ] Performance seems worse somehow?
- [ ] When pressing `q`, Qt 5 qutebrowser interprets the next keypress as macro register and goes back to normal mode. With Qt 6, it gets stuck in `record_macro` mode without any visual indication of that being the case, so keypresses don't work until pressing esc.

**Comment by bodograumann** [comment]
Hit a bug due to outdated chromium version: https://github.com/tailwindlabs/tailwindcss.com/issues/1122

Will chromium be updated in qt5-webengine at some point, or will we have to wait until this upgrade to qt6 is complete?

**Comment by The-Compiler** [comment]
> Will chromium be updated in qt5-webengine at some point, or will we have to wait until this upgrade to qt6 is complete?

You will need to wait for this to land, unfortunately (or use the `qt6-test` branch already if you don't mind some random bugs).

**Comment by tinywrkb** [comment]
[Flatpak test builds](https://github.com/flathub/org.qutebrowser.qutebrowser/pull/138)

The 6.2 branch of the QtWebEngine Flatpak base app is now available to install from Flathub, so we can start generating PyQt6 test builds of qutebrowser.

**Comment by toofar** [comment]
Quick update from on the 6.3 branch for the last few days. Some things are fixed (I've checked them up above too:

- [x] stylesheet issues (mostly visible in the completion widget not picking up our styles) appear to be fixed, everything is back to normal
- [x] similarly the styling issues with the context menu (text and backround the same color, fixed width) are resolved (although i'm glad it pushed me to style the context menu)
- [x] google meet works again
- [x] "select element" from the web inspector highlights page elements when you hover them again

Some things from @The-Compiler's last post with checkboxes I can't reproduce:

> :report test

seems to work for me (at least it told me to upgrade from 2.3.something to 2.4). I sent one in about 30m ago (2022-02-27T04:15Z or so)

> Sometimes my mouse wheel seems to break

Hmmm, not sure if I've seen that. Sometimes input into my whole X session breaks until I do something like switch VTs and kill something, not sure if it has anything to do with qutebrowser.

> Performance seems worse somehow? And white squares on dark pages.

I can't say I've noticed anything myself. Although 6.2.2 -> 6.3 seems to use way more memory. My binaries aren't stripped on either version so I'm not 100% sure I'm building them right (libQt6WebEngineCore.so is 240M on 6.2.2 and 260M on 6.3 so probably that isn't the reason for any difference).

I can reproduce everything else in that long list. (You can see the insecure download abort crash on https://www.technic3d.com/Knoppix-Mirror/knoppix-dvd/ too.)

Another one I'm seeing on this build, although since it isn't coming from our code it probably comes from building on a branch where all the submodules aren't synchronoised for a release:

```
  # Trying to choose a file to upload to confluence (QWebEnginePage.chooseFiles()), also :print (QPrintDialog())
  # I got a stack trace by defining a `QStyle.PixelMetric._missing_` method that dropped into a pdb shell
  Traceback (most recent call last):
  File "/usr/lib/python3.9/enum.py", line 384, in __call__
    return cls.__new__(cls, value)
  File "/usr/lib/python3.9/enum.py", line 702, in __new__
    raise ve_exc
  ValueError: 94 is not a valid QStyle.PixelMetric
```

In summary it is shaping up pretty good from the Qt side. There are issues with favicons, segfaults in the tests, pyinstaller issues and we get worse stack traces from PyQt. Otherwise it's looking like it's all on us to adapt to a few deprecations and fix up lots and lots of cases where we deal with enums in ways that no longer work.

**Comment by Bonoolu** [comment]
Hi, i switched to this branch because of the recent CVE. Were you able to solve the white square issue? Is there a workaround?

**Comment by foxxx0** [comment]
Gave this branch a quick test (very brief) in its current state but it already looks quite decent.

Though one thing looked off to me in `qute://version`:
```
PyQt6.QtWebEngine: no
```
Even though I had `python-pyqt6` and `python-pyqt6-webengine` installed (alongside `qt6-base` and `qt6-webengine` obviously), similar to the screenshot ( https://github.com/qutebrowser/qutebrowser/issues/5395#issuecomment-1002264243 ) by @tinywrkb  from a bit earlier.
The `backend` however is correctly shown as `QtWebEngine 6.2.x`.

I have now taken a couple of minutes and rebased this `qt6-test` branch onto the current `master`, which can now be found at https://github.com/foxxx0/qutebrowser/tree/qt6-test-rebased .
(This was just a bit mangling stuff and I have likely purged too much of the tox/requirements stuff, but at least it's something.)
It wasn't really that much effort and didn't uncover any major headaches, at least none that I stumbled upon.
I just went with `git+https://github.com/TilmanK/PyQt6-stubs@main#egg=PyQt6-stubs` in `misc/requirements/requirements-mypy.txt-raw` for now, this will likely need adjusting once you sort out who/where the final stubs package is being developed.


Until now, I have not stumbled upon the bugs that @Bonoolu commited here: https://github.com/Bonoolu/qutebrowser/commit/b5084f9dc7604b74237881ea476ed488174b4ead , will post an update if that changes.


As the just released 2.5.0 marks the last 2.x release, I think it would be a good time to rebase the "official" `qt6-test` branch onto `2.5.0`/`master` as well. :)

**Comment by The-Compiler** [comment]
I do plan working on a "real" Qt 6 branch next, now that 2.5.0 is out. I will probably recreate it rather than rebasing though, as most of the conflicting changes are automated, and some of the manually done changes were hacks which should be solved in a better way now or aren't needed anymore.

**Comment by The-Compiler** [comment]
> Were you able to solve the white square issue? Is there a workaround?

So you see it as well? So far nobody else did... what kind of (graphics) hardware is this?

I'm not aware of any workaround.

**Comment by Bonoolu** [comment]
> So you see it as well? So far nobody else did... what kind of (graphics) hardware is this?
> 
> I'm not aware of any workaround.

Yes i see it as well. I have an AMD Vega 56. The white square is jumping around when i move my mouse.
Here are a few pics of the square jumping around.
The last pic looks interesting. the white square is rendering between different layers.
![1](https://user-images.githubusercontent.com/64061670/161440491-af0a1bb0-0bd7-4123-ab1b-0110efcc346a.jpg)
![2](https://user-images.githubusercontent.com/64061670/161440492-abd32ac6-df71-42f7-aaeb-5a4353327b54.jpg)
![3](https://user-images.githubusercontent.com/64061670/161440493-269346ae-fc77-4009-84ec-73a2a9164b8e.jpg)

**Comment by Bonoolu** [comment]
The white square stops jumping and freezes in place when i open the context menu. It stays until i close the context menu and then move my mouse

**Comment by The-Compiler** [comment]
> I have an AMD Vega 56.

That's rather suspicious - I use the internal graphics of an AMD Ryzen 7 PRO 4750U.

Anyone else on AMD graphics who can reproduce this?

@Bonoolu Any chance you could try with the Qt simplebrowser example? You can get it e.g. by cloning https://code.qt.io/qt/qtwebengine.git, doing `git checkout 6.2.4`` (assuming that's the Qt you have), going to `qtwebengine/examples/webenginewidgets/simplebrowser/`, then running `cmake .`, `make -j$(nproc)` and finally `./simplebrowser`.

For some reason, I could never reproduce it there, but I don't know what would cause it in qutebrowser.

**Comment by toofar** [comment]
I got a BLUE square today! When opening the github notifications page in a new tab (I've got process-model=process-per-site and I have other github tabs open), using the github dark userscript. It appeared when I hovered over one of the list items near it, disappeared when I didn't. It went away for good before I could get a screenshot. I haven't changed anything with my setup, first time I've got it. AMD 6700XT. Also various maybe-hardware-acceleration-related flags in qt.args. Could not reproduce with simplebrowser (but this is the first time I've produced it with qutebrowser so that's not saying much).

**Comment by Bonoolu** [comment]
> For some reason, I could never reproduce it there, but I don't know what would cause it in qutebrowser.
I also can't reproduce it in simplebrowser.

Sometimes the white square overlays the statusbar at the bottom of qutebrowser, making the url unreadable.
i mean this bar:
![image](https://user-images.githubusercontent.com/64061670/161594001-07628541-2623-4055-9f89-80b714cd584b.png)

**Comment by tinywrkb** [comment]
With the recent key events handling changes, the Flatpak builds are actually usable. It was crashing before this on modifier key combinations, and now it's pretty stable.

A minor regression is that qutebrowser gets a `org.mpris.MediaPlayer2.chromium.instance[0-9]+` D-Bus name for its MPRIS interface, e.g. `org.mpris.MediaPlayer2.chromium.instance124`, instead of `org.mpris.MediaPlayer2.qutebrowser.instance[0-9]+`.  
This seems to be a regression Chromium as I noticed this also in Chrome deb releases, stable and unstable, and only Edge have this set correctly, using `org.mpris.MediaPlayer2.edge.instance[0-9]+`.  
I'm actually stopped using MPRIS, so I'm just looking the xdg-dbus-proxy log (`$ flatpak run --log-session-bus ...`) and D-Feet.

**Comment by The-Compiler** [comment]
> This seems to be a regression Chromium as I noticed this also in Chrome deb releases, stable and unstable, and only Edge have this set correctly, using `org.mpris.MediaPlayer2.edge.instance[0-9]+`.

Don't think so - Chromium actually [hardcodes](https://source.chromium.org/chromium/chromium/src/+/main:components/system_media_controls/linux/system_media_controls_linux.cc;l=53-54;drc=88b2443b34a9c231906956eaf4888cee2062105c) this unfortunately, and QtWebEngine [used to patch it](https://codereview.qt-project.org/c/qt/qtwebengine-chromium/+/307217). Maybe that patch got lost while they were rebasing things. Will open an issue once I go through the comments here again.

**Comment by tinywrkb** [comment]
> Don't think so - Chromium actually [hardcodes](https://source.chromium.org/chromium/chromium/src/+/main:components/system_media_controls/linux/system_media_controls_linux.cc;l=53-54;drc=88b2443b34a9c231906956eaf4888cee2062105c) this unfortunately, and QtWebEngine [used to patch it](https://codereview.qt-project.org/c/qt/qtwebengine-chromium/+/307217).

Well, then maybe Google was also patching this ;), because I'm pretty sure the Chrome deb releases were using `org.mpris.MediaPlayer2.chrome.*` as I've been maintaining my own private Chrome stable and unstable Flatpaks and I poked a hole in the sandbox for this D-Bus name ownership.

**Comment by tinywrkb** [comment]
Today's experiment.

![swappy-20220407-143705](https://user-images.githubusercontent.com/44030034/162190503-a2aa4312-7fd5-4e53-ae4c-a81080e4678d.png)

**Comment by Bonoolu** [comment]
I sometimes get Segmentation faults when i close a tab with the middle mouse button or when i close a browser window.
i would make a stack trace, but i can't build qt6-webengine 6.2.4 with debug symbols.
My pc runs out of memory. i have 16GB ram and 16GB swap.

**Comment by mschilli87** [comment]
> My pc runs out of memory. i have 16GB ram and 16GB swap.

You might be able to trade memory for build time by reducing the number of parallel Nake jobs and/or disabling Unity/Jumbo builds.

**Comment by tinywrkb** [comment]
Should we switch to build from `qt6-v2`?  
If there's interest, and there's a green light from the developers here, then I can set up a beta branch for publishing via the flathub-beta repo, though it will stay on QtWebEngine 6.2.x until there's actual KDE 6.3 runtime, which I have no idea yet if it will be synced with Freedesktop 22.08 (yeh, Aug-2022).

**Comment by The-Compiler** [comment]
Hehehe, I suspected I couldn't sneakily push some work-in-progress without someone noticing it :sweat_smile: 

It's not ready yet - the idea of `qt6-v2` is to have a new updated branch with Qt 6 support, but with less hacks, while keeping Qt 5 support around, and with a cleaner history. Right now it's in the state that `qt6-test` was back in August or so: It runs, but as soon as you press a key, it will crash.

I'll cherry-pick the existing fixes in `qt6-test` over the next couple of days and let you know once it's ready to use.

**Comment by The-Compiler** [comment]
Quick update: As of Sunday or so, `qt6-v2` should be more stable than `qt6-test` is, and the latter should be regarded as abandoned (probably gonna rename it to make that clearer).

Still working out some kinks and will probably force-push `qt6-v2` from time to time to clean up the history, but it should be preferred to `qt6-test` as of now.

**Comment by bodograumann** [comment]
Is it possible to run both versions of qutebrowser in parallel – something like `--new-instance` in firefox?

**Comment by The-Compiler** [comment]
@bodograumann Use a different `--basedir`, or use `--temp-basedir`.

**Comment by The-Compiler** [comment]
I've made some progress, and the Qt 5 (!) tests now pass on `qt6-v2`. I've also dropped Qt < 5.15 because I was tired of adding compatibility code which will get deleted again in a few weeks.

Next up over the next couple of weeks:

- [ ] Look at the ~120 remaining test failures with Qt 6.2
- [ ] Check whether we can get static analysis (and e.g. VS Code completion) to work somehow with our `from ... import *` approach - what do `qts` or `qtpy` do?
    - [ ] Calibre [ships with](https://github.com/kovidgoyal/calibre/tree/master/src/qt) `.pyi` files listing every Qt member - I don't like that idea.
- [ ] Look at Qt 6.3
- [ ] Get `mkvenv.py` to work with `--pyqt-version 6.3`
- [ ] Clean up this issue
    - [ ] split up any remaining problems into their own issues
    - [ ] add some instructions to the first post
    - [ ] read through everything again and see if we missed any API changes
- [ ] Maybe get linters to work (by making them e.g. ignore long lines temporarily)
- [ ] Probably merge some PRs while waiting for more feedback and to avoid crazy conflicts with the next steps
- [ ] Integrate #7128 if we didn't find a way to make static analysis work otherwise
- [ ] Adopt `black` (#1455) so that we pass "normal" linting again and don't have crazy long lines

**Comment by The-Compiler** [comment]
Two more issues I noticed:

- [ ] When opening an upload dialog (e.g. clicking the status bar here on GitHub):

```pytb
Traceback (most recent call last):
  File "/usr/lib/python3.10/enum.py", line 385, in __call__
    return cls.__new__(cls, value)
  File "/usr/lib/python3.10/enum.py", line 710, in __new__
    raise ve_exc
ValueError: 94 is not a valid QStyle.PixelMetric
```

- [ ] Our workaround for #5182 (Google sign in) does not seem to work anymore. Maybe this is the end of it? Don't believe there is much we can do if it really doesn't work anymore, it's not like Google will care about us.

**Comment by quotquot** [comment]
FYI, with https://salsa.debian.org/python-team/packages/pyqt6/ and https://salsa.debian.org/python-team/packages/pyqt6-webengine/, qt6-v2 now runs natively on Debian sid/testing:

```
qutebrowser v2.5.0
Git commit: de99a57 on qt6-v2 (2022-04-22 12:07:41 +0200)
Backend: QtWebEngine 6.2.4, based on Chromium 90.0.4430.228
Qt: 6.2.4

CPython: 3.10.4
PyQt: 6.3.0

sip: 4.19.25
colorama: 0.4.4
jinja2: 3.0.3
pygments: 2.11.2
yaml: 5.4.1
adblock: no
PyQt5.QtWebEngineWidgets: yes
PyQt5.QtWebEngine: 5.15.5
PyQt5.QtWebKitWidgets: no
pdf.js: 2.6.0 (/usr/share/javascript/pdf/build/pdf.js)
sqlite: 3.38.2
QtNetwork SSL: OpenSSL 1.1.1n  15 Mar 2022

Style: QFusionStyle
Platform plugin: wayland
OpenGL: Intel, 3.1 Mesa 22.0.2
Platform: Linux-5.17.0-1-amd64-x86_64-with-glibc2.34, 64bit
Linux distribution: Debian GNU/Linux bookworm/sid (debian)
```

**Comment by The-Compiler** [comment]
Quick update: The tests in `tests/unit/` all pass now, with Qt 5.15, 6.2 and 6.3.

Next up, `tests/end2end` - current state there:

- 0 failures with Qt 5.15
- 18 failures with Qt 6.2
- 22 failures with Qt 6.3

Some issues on Qt 6, from what I can see:

- [ ] Binding keys which are not listed in the [`Qt::Key` enum](https://doc.qt.io/qt-6/qt.html#Key-enum) can not be bound (not sure if they work in insert mode either). This is mostly stuff outside of ASCII/Latin1, e.g. cyrillic or greek letters. It's unlikely that qutebrowser can fix this [without PyQt's help](https://www.riverbankcomputing.com/pipermail/pyqt/2022-April/044607.html).
- [ ] The `content.headers.referer` setting (i.e. only sending the referer for the same domain) might not work yet.
- [ ] Some dark mode stuff might not work yet.
- [ ] Something about redirects in downloads might be broken.
- [ ] Something about recording/replaying keyboard macros might be broken.
- [ ] `tabs.mode_on_change` might not work yet.

If you want to try reproducing some of those and describe the difference you can see between Qt 5 and 6 (can set `QUTE_QT_WRAPPER=PyQt5` in the environment, make sure to use `--temp-basedir`), that'd be helpful! But I hope to see what's going on from looking at the tests too.

**Comment by chadkouse** [comment]
Hi I've been using the `qt6-v2` branch for a few days now and it's been pretty stable and good for me... I did notice that enough code changed where that "better input detection" patch doesn't apply anymore and I can't even see how I could manually hack it in.  The KeyInput EventFilter seems to have changed substantially.  I'd like to rework this patch but I really don't know where to start, any pointers?

**Comment by The-Compiler** [comment]
> Hi I've been using the `qt6-v2` branch for a few days now and it's been pretty stable and good for me...

Good to hear!

> I did notice that enough code changed where that "better input detection" patch doesn't apply anymore

"that 'better input detection' patch"?

> and I can't even see how I could manually hack it in. The KeyInput EventFilter seems to have changed substantially.

`qutebrowser/keyinput/eventfilter.py` had no changes (other than enum member access).

> I'd like to rework this patch but I really don't know where to start, any pointers?

You'll need to be more specific about what patch you mean and which changes are giving you trouble - right now I don't really follow I'm afraid.

**Comment by Bonoolu** [comment]
qutebrower crashed for me when i tried to open a website with an expired certificate

```
18:50:26 ERROR    network    webenginetab:_on_ssl_errors:1569 Non-overridable certificate error: Server's certificate has expired.
18:50:26 DEBUG    network    webenginetab:_on_ssl_errors:1572 ignore False, URL PyQt6.QtCore.QUrl('https://www.techcoil.com/blog/how-to-save-and-load-objects-to-and-from-file-in-python-via-facilities-from-the-pickle-module/'), requested PyQt6.QtCore.QUrl('https://www.google.com/search?hl=en&q=write python object to file')
18:50:26 ERROR    misc       crashsignal:_handle_early_exits:241 Uncaught exception
Traceback (most recent call last):
  File "/usr/lib/python3.10/site-packages/qutebrowser/browser/webengine/webview.py", line 192, in _handle_certificate_error
    return error.certificate_was_accepted()
  File "/usr/lib/python3.10/site-packages/qutebrowser/utils/usertypes.py", line 519, in certificate_was_accepted
    raise ValueError("No decision taken yet")
ValueError: No decision taken yet
```

**Comment by The-Compiler** [comment]
> qutebrower crashed for me when i tried to open a website with an expired certificate

Reproducable if a page uses HSTS only. Reproduced with: `qutebrowser --debug --temp-basedir 'https://preloaded-hsts.badssl.com/' ':later 1000 open https://subdomain.preloaded-hsts.badssl.com/'`

**Comment by The-Compiler** [comment]
With 7508165fd1b1ece2eef0608cd548b628dccf0f3d, `mkvenv.py` now works properly again. Updated instructions for trying the `qt6-v2` branch:

- **Take a backup of your qutebrowser data directory** (see `:version`) unless you don't mind losing cookies and such when down-/upgrading. This isn't needed if you use `--temp-basedir` below, but if you forget to do so, your data will get upgraded, and as far as I'm aware, there is no way to read it with Qt 5 after that.
- Clone the git repo, e.g. via `git clone https://github.com/qutebrowser/qutebrowser.git`
- `cd qutebrowser`
- Check out the branch using `git checkout qt6-v2`
- Create a new separate environment via `python3 scripts/mkvenv.py --venv-dir .venv-qt6 --pyqt-version 6.3`
- Run `.venv-qt6/bin/qutebrowser --temp-basedir` and verify via `:version` that you're indeed on Qt 6

---

@Bonoolu Should now be fixed too in 7508165fd1b1ece2eef0608cd548b628dccf0f3d.

**Comment by The-Compiler** [comment]
Note to self, new darkmode feature with QtWebEngine 6.3: [Add dark mode feature for increasing text contrast (I5e8e0639) · Gerrit Code Review](https://chromium-review.googlesource.com/c/chromium/src/+/2893236) (92.0.4507.0)

**Comment by The-Compiler** [comment]
And another note to self, failing darkmode tests:

# test_preferred_colorscheme_with_dark_mode

<details>

## 5.15

![screenshot-0](https://user-images.githubusercontent.com/625793/167271634-feca425c-5514-4e06-ad80-57ad8ddb14b4.png)

## 6.2

![screenshot-4](https://user-images.githubusercontent.com/625793/167271641-fd9e562a-1ada-43a7-af99-e06ffed90dc3.png)

</details>

# test_dark_mode[blank-lightness-cielab-colors0]

<details>

## 5.15

![screenshot-0](https://user-images.githubusercontent.com/625793/167271669-c2c129fc-87a0-463e-910b-4c68616de952.png)

## 6.2

![screenshot-4](https://user-images.githubusercontent.com/625793/167271676-1e12fb8e-8f5b-4e23-b7a0-1bbf60873c39.png)

</details>

# test_dark_mode[yellow-lightness-cielab-colors3] 

<details>

## 5.15

![screenshot-1](https://user-images.githubusercontent.com/625793/167271696-fb6a3a42-e3a4-4d13-9454-4167a459d6f4.png)

## 6.2

![screenshot-4](https://user-images.githubusercontent.com/625793/167271691-eaf0135a-7258-4a84-9d7c-c871123d6d37.png)

</details>

# test_dark_mode[yellow-lightness-hsl-colors4]

<details>

## 5.15

![screenshot-1](https://user-images.githubusercontent.com/625793/167271717-40c4718f-9eb3-4fb1-b670-8ba34a584410.png)

## 6.2

![screenshot-4](https://user-images.githubusercontent.com/625793/167271711-da19571d-15b9-4578-80b1-ea288f80c8bd.png)

</details>

# test_dark_mode[yellow-brightness-rgb-colors5]

<details>

## 5.15

![screenshot-1](https://user-images.githubusercontent.com/625793/167271746-ff155166-707d-4906-afc2-b40342ad490e.png)

## 6.2

![screenshot-4](https://user-images.githubusercontent.com/625793/167271735-84c1c9ee-1add-45a8-a95e-fd7f8952b74e.png)

</details>

# test_dark_mode_mathml

<details>

## 5.15

![screenshot-1](https://user-images.githubusercontent.com/625793/167271765-4df37bdd-6b65-4973-912f-a6846e61a58c.png)

## 6.2

![screenshot-4](https://user-images.githubusercontent.com/625793/167271754-cc4bc623-580f-4b38-bb71-de993d3746fb.png)

</details>

Most of them look like minor changes in Chromium's algorithm somewhere.

**Comment by The-Compiler** [comment]
One important milestone reached: The testsuite passes with Qt 6.2 and 6.3 :tada: :sparkles: 

![image](https://user-images.githubusercontent.com/625793/167274653-531de3aa-ae1c-4480-bdcc-3e76fd8e6f45.png)

**Comment by The-Compiler** [comment]
@tinywrkb Any chance you could run the testsuite on Qt 6.3 (or 6.2) with aarch64 and get me a log of the results easily? I suspect it will need some adjustments for expected dark mode colors again, but I'm not sure what the values will be exactly. If not, no worries, then I'll get my Raspberry Pi out again.

**Comment by tinywrkb** [comment]
We don't have a runtime with Qt 6.3. KDE 6.3 runtime will likely be released when the Freedesktop 22.08 will be available, so we have to wait ~3 months.

Tests are now running (still ongoing) with Qt 6.2 on the Flathub CI: https://flathub.org/builds/#/builders/32/builds/90326

**Comment by tinywrkb** [comment]
Branch `qt6-v2` built with KDE 6.2 runtime.

**aarch64** 

> 55 failed, 9055 passed, 405 skipped, 62 xfailed, 61 errors, 6 rerun in 1749.74s (0:29:09)

[Full build log](https://flathub.org/builds/api/v2/logs/3611320/raw)

**x86_64**

> 51 failed, 9058 passed, 405 skipped, 63 xfailed, 64 errors, 6 rerun in 1368.99s (0:22:48)

[Full build log](https://flathub.org/builds/api/v2/logs/3611319/raw)

**Comment by The-Compiler** [comment]
@tinywrkb Thanks! Most of the errors seem to be due to https://github.com/pytest-dev/pytest-bdd/issues/447 - could you perhaps rerun with `pytest-bdd<5`?

**Comment by tinywrkb** [comment]
With `pytest-bdd` downgraded to `4.1.0`.

**aarch64**

> 17 failed, 9093 passed, 406 skipped, 62 xfailed, 62 errors, 1 rerun in 1782.96s (0:29:42)

[Build log](https://flathub.org/builds/api/v2/logs/3611541/raw)

**x86_64**

> 13 failed, 9096 passed, 406 skipped, 63 xfailed, 65 errors, 1 rerun in 1367.24s (0:22:47)

[Build log](https://flathub.org/builds/api/v2/logs/3611540/raw)

**Comment by The-Compiler** [comment]
@tinywrkb One error I see a lot now is "QStandardPaths: wrong permissions on runtime directory /run/user/995, 0755 instead of 0700". I could ignore it in the tests, but it sounds like this is something that should be fixed in the environment you are running the tests in instead?

**Comment by tinywrkb** [comment]
That's Flathub's CI. Last time I asked, this part of Flathub's infrastructure doesn't have its configuration publicly documented and have sources/recipes on some public repo.  
I will try to chmod this folder before running the tests, not sure if it will work though.

**Comment by tinywrkb** [comment]
Changing `XDG_RUNTIME_DIR`'s mode bits works.

**aarch64**

>  8 failed, 9101 passed, 406 skipped, 62 xfailed in 1790.48s (0:29:50)

[Build log](https://flathub.org/builds/api/v2/logs/3612189/raw)

**x86_64**

>  3 failed, 9105 passed, 406 skipped, 63 xfailed in 1356.09s (0:22:36)

[Build log](https://flathub.org/builds/api/v2/logs/3612188/raw)

**Comment by The-Compiler** [comment]
Cool, thanks again! I see four different issues:

- The one I was expecting, slightly different color values on ARM. Should be fixed in 2f44ed2e83720539ec2b112bf1f5326abeff1002.
- Flakiness in `tests/end2end/features/test_navigate_bdd.py::test_navigating_up_in_qutehelp`. Could reproduce locally, and 476fd65b2736310f6451bb59a3f3a8421808d0c1 should fix it.
- Flakiness in `tests/end2end/features/test_editor_bdd.py::test_spawning_an_editor_and_saving`. Can't reproduce, but I think I saw it on CI before too. Not sure what to do about that one, but I expect it to happen in `master` too.
- Missing sandboxing from Chromium (probably will fail on `master` too). I opened https://github.com/flathub/org.qutebrowser.qutebrowser/issues/193 for it.

Could you do another rerun to make sure I didn't do some stupid mistake? If that looks better, I'd like to next proceed with removing lots of old ignored error/warning messages. Then I'll probably ask you for another rerun, since some of those might still needed in the flatpak environment.

**Comment by tinywrkb** [comment]
[Running...](https://flathub.org/builds/#/builders/32/builds/90471)

**Comment by tinywrkb** [comment]
**aarch64**

> 2 failed, 9106 passed, 406 skipped, 63 xfailed in 1738.08s (0:28:58)

[Build log](https://flathub.org/builds/api/v2/logs/3616424/raw)

**x86_64**

> 2 failed, 9106 passed, 406 skipped, 63 xfailed in 1345.00s (0:22:25)

[Build log](https://flathub.org/builds/api/v2/logs/3616423/raw)

**Comment by The-Compiler** [comment]
Note to self:

\<saro\> The-Compiler is it expected `c.qt.force_software_rendering = 'qt-quick'` to crash qt6-v2 ?

**Comment by The-Compiler** [comment]
@tinywrkb This looks great! Only the sandboxing tests remaining, which I adjusted in a19175b23d73252be2a89d9781ff0c87d344e306. Will adjust them again if/when https://github.com/flathub/io.qt.qtwebengine.BaseApp/pull/66 gets in. 

With 3bc71bb0c5acb34138007a3ddef5ccfa333858fd I removed all the former log ignores (with a couple readded in follow-up commits) - do you see any regressions after that? Any chance you could run the latest `qt6-v2` branch but with Qt 5 as well? (should be auto-detected as long as Qt 6 is unavailable - if that's not an option, set `QUTE_QT_WRAPPER=PyQt5` in the environment).

Thanks for all your work! :sparkles:

**Comment by tinywrkb** [comment]
### PyQt6

**aarch64**

> 11 failed, 9097 passed, 407 skipped, 63 xfailed, 66 errors, 1 rerun in 1772.03s (0:29:32)

[Build log](https://flathub.org/builds/#/builders/51/builds/6999/steps/6/logs/stdio)

**x86_64**

> 12 failed, 9096 passed, 407 skipped, 63 xfailed, 66 errors, 1 rerun in 1339.96s (0:22:19)

[Build log](https://flathub.org/builds/api/v2/logs/3629108/raw)

### PyQt5

**aarch64**

> 2 failed, 9125 passed, 384 skipped, 66 xfailed, 2 errors in 1724.93s (0:28:44)

[Build log](https://flathub.org/builds/api/v2/logs/3629133/raw)

**x86_64**

> 2 failed, 9125 passed, 384 skipped, 66 xfailed, 2 errors in 1271.47s (0:21:11)

[Build log](https://flathub.org/builds/api/v2/logs/3629132/raw)

**Comment by tinywrkb** [comment]
It looks like I forgot the `XDG_RUNTIME_DIR` mode bits fix when I rebased both branches, PyQt6 and PyQt5.

**Comment by tinywrkb** [comment]
### PyQt6

**aarch64**

> 3 failed, 9104 passed, 407 skipped, 63 xfailed, 2 errors in 1745.55s (0:29:05)

[Build log](https://flathub.org/builds/api/v2/logs/3629289/raw)

**x86_64**

> 2 failed, 9105 passed, 407 skipped, 63 xfailed, 2 errors in 1308.67s (0:21:48)

[Build log](https://flathub.org/builds/api/v2/logs/3629268/raw)

### PyQt5

**aarch64**

> 3 failed, 9124 passed, 384 skipped, 66 xfailed, 2 errors in 1733.36s (0:28:53)

[Build log](https://flathub.org/builds/api/v2/logs/3629290/raw)

**x86_64**

> 2 failed, 9125 passed, 384 skipped, 66 xfailed, 2 errors in 1250.93s (0:20:50)

[Build log](https://flathub.org/builds/api/v2/logs/3629288/raw)

**Comment by The-Compiler** [comment]
@tinywrkb Yay, so only one failure because of unexpected error messages. Fixed that in 6fae468b6f83d35557ed216ac104ec3e79f39442 and the rest in 6b208a61602b56249aa535bfaf96d9025e7de5be and be441b46aadbd3cb001a30993a7867146ece4494.

Two hypothesis errors are a wontfix, probably just caused by the CI environment:

> Unreliable test timings! On an initial run, this test took 1552.28ms, which exceeded the deadline of 600.00ms, but on a subsequent run it took 0.37 ms, which did not. If you expect this sort of variability in your test timings, consider turning deadlines off for this test by setting deadline=None.
>
> Unreliable test timings! On an initial run, this test took 1344.36ms, which exceeded the deadline of 600.00ms, but on a subsequent run it took 0.48 ms, which did not. If you expect this sort of variability in your test timings, consider turning deadlines off for this test by setting deadline=None.

Could you rerun to confirm the issues are gone now?

**Comment by tinywrkb** [comment]
### PyQt6

**aarch64**

> =========================== short test summary info ============================
   FAILED tests/end2end/test_invocations.py::test_sandboxing[enable-all-True-True-True-You are adequately sandboxed.]
   FAILED tests/end2end/test_invocations.py::test_sandboxing[disable-seccomp-bpf-True-False-True-You are NOT adequately sandboxed.]
   ===== 2 failed, 9106 passed, 406 skipped, 63 xfailed in 1849.89s (0:30:49) =====

[Build log](https://flathub.org/builds/api/v2/logs/3634038/raw)

**x86_64**

> =========================== short test summary info ============================
   FAILED tests/end2end/test_invocations.py::test_sandboxing[enable-all-True-True-True-You are adequately sandboxed.]
   FAILED tests/end2end/test_invocations.py::test_sandboxing[disable-seccomp-bpf-True-False-True-You are NOT adequately sandboxed.]
   ===== 2 failed, 9106 passed, 406 skipped, 63 xfailed in 1512.36s (0:25:12) =====

[Build log](https://flathub.org/builds/api/v2/logs/3634031/raw)

### PyQ5

**aarch64**

> =========================== short test summary info ============================
   FAILED tests/end2end/test_invocations.py::test_sandboxing[enable-all-True-True-True-You are adequately sandboxed.]
   FAILED tests/end2end/test_invocations.py::test_sandboxing[disable-seccomp-bpf-True-False-True-You are NOT adequately sandboxed.]
   ===== 2 failed, 9126 passed, 383 skipped, 66 xfailed in 1822.71s (0:30:22) =====

[Build log](https://flathub.org/builds/api/v2/logs/3634113/raw)

**x86_64**

> =========================== short test summary info ============================
   FAILED tests/end2end/test_invocations.py::test_sandboxing[enable-all-True-True-True-You are adequately sandboxed.]
   FAILED tests/end2end/test_invocations.py::test_sandboxing[disable-seccomp-bpf-True-False-True-You are NOT adequately sandboxed.]
   ===== 2 failed, 9126 passed, 383 skipped, 66 xfailed in 1291.22s (0:21:31) =====

[Build log](https://flathub.org/builds/api/v2/logs/3634096/raw)

**Comment by The-Compiler** [comment]
@tinywrkb I just pushed another fix, could you give it a spin? Fingers crossed that it will work this time around....

---

Also, quick status update: `qt6-v2` is coming along nicely, and from an user's perspective, should be fine to test now really. I *think* I fixed the last known crashes (the last one was when I was trying to print a website, guess nobody tried that so far :rofl:). Next up is a rebase of the >200 commits in the branch to end up with a nice overview of what was actually changed why, and then we "only" need to figure out how to make it fit all together (#7128, some other major open PRs I'd really like to merge, reformatting everything, etc.)...

**Comment by tinywrkb** [comment]
> @tinywrkb I just pushed another fix, could you give it a spin? Fingers crossed that it will work this time around....

- [PyQt5](https://github.com/flathub/org.qutebrowser.qutebrowser/pull/199)
- [PyQt6](https://github.com/flathub/org.qutebrowser.qutebrowser/pull/200)

**Comment by tinywrkb** [comment]
### PyQt6

**aarch64**

> =========================== short test summary info ============================
   FAILED tests/end2end/test_invocations.py::test_sandboxing[disable-all-False-False-False-You are NOT adequately sandboxed.]
  ===== 1 failed, 9133 passed, 406 skipped, 63 xfailed in 1956.65s (0:32:36) =====

[Build log](https://flathub.org/builds/api/v2/logs/3669069/raw)

**x86_64**

> =========================== short test summary info ============================
   FAILED tests/end2end/test_invocations.py::test_sandboxing[disable-all-False-False-False-You are NOT adequately sandboxed.]
   ===== 1 failed, 9133 passed, 406 skipped, 63 xfailed in 1546.01s (0:25:46) =====

[Build log](https://flathub.org/builds/api/v2/logs/3669067/raw)

### PyQt5

**aarch64**

> =========================== short test summary info ============================
   FAILED tests/end2end/test_invocations.py::test_sandboxing[disable-all-False-False-False-You are NOT adequately sandboxed.]
   FAILED tests/unit/utils/test_urlutils.py::TestProxyFromUrl::test_proxy_from_url_pac[pac+http]
   ===== 2 failed, 9152 passed, 383 skipped, 66 xfailed in 1753.26s (0:29:13) =====

[Build log](https://flathub.org/builds/api/v2/logs/3669068/raw)

**x86_64**

> =========================== short test summary info ============================
   FAILED tests/end2end/test_invocations.py::test_sandboxing[disable-all-False-False-False-You are NOT adequately sandboxed.]
   ===== 1 failed, 9153 passed, 383 skipped, 66 xfailed in 1338.13s (0:22:18) =====

[Build log](https://flathub.org/builds/api/v2/logs/3669062/raw)

**Comment by The-Compiler** [comment]
@tinywrkb Aaargh... ok, one more time please, I think I got it figured out now.

Also, it'd be good if you could remove pytest-icdiff (ref 81f0b292912f26afb21f979ae2f45ef8e067e679) - unfortunately, in situations like this, it makes it *more* difficult to see where the culprit is, which is kind of the opposite of what it was intended for.

**Comment by tinywrkb** [comment]
### PyQt6

**aarch64**

> 9132 passed, 408 skipped, 63 xfailed in 1814.68s (0:30:14)

[Build log](https://flathub.org/builds/api/v2/logs/3669575/raw)

**x86_64**

> 9132 passed, 408 skipped, 63 xfailed in 1541.56s (0:25:41)

[Build log](https://flathub.org/builds/api/v2/logs/3669567/raw)

### PyQt5

**aarch64**

> 9152 passed, 385 skipped, 66 xfailed in 2116.93s (0:35:16)

[Build log](https://flathub.org/builds/api/v2/logs/3669576/raw)

**x86_64**

> 9152 passed, 385 skipped, 66 xfailed in 1480.15s (0:24:40)

[Build log](https://flathub.org/builds/api/v2/logs/3669574/raw)

**Comment by The-Compiler** [comment]
Yesss, finally! Thanks @tinywrkb for triggering all those builds for me. Hope I won't need another ~5 tries once https://github.com/flathub/io.qt.qtwebengine.BaseApp/pull/66 is in :sweat_smile: 

FYI, planning to go through this issue and hide a lot of older comments soon, as a lot of stuff is outdated/resolved by now.

**Comment by tinywrkb** [comment]
No problem!  
[KDE 6.3 runtime](https://invent.kde.org/packaging/flatpak-kde-runtime/-/tree/qt6.3) should be available soon, so we will have to run the tests again with PyQt6.

I took the liberty to publish beta builds with PyQt6 through the flathub-beta repo.  
It's QtWebEngine 6.2 of course, so it will be a bit more interesting when the KDE 6.3 runtime will be out.

The beta branch uses the same app-id as the stable channel (master branch), so the warning about data dir applies here.  
I don't see a way to change datadir. If there's one, then I can add a wrapper to set it to something like `$XDG_DATA_HOME/qutebrowser-beta`, if it's even something that we want.

No automated "nightly" rebuilds yet, this is being updated manually. I'll try writing a GitHub workflow for this. 

Adding the flathub-beta remote and installing from the beta branch can be done through an application manager (Gnome Software, Discover, ...) or via the command-line.

```
$ flatpak remote-add --if-not-exists flathub-beta https://flathub.org/beta-repo/flathub-beta.flatpakrepo
$ flatpak install flathub-beta org.qutebrowser.qutebrowser//beta
```

**Comment by tinywrkb** [comment]
Overriding `XDG_DATA_HOME` is also not something I want to do. I'll just leave it as it is.   
It would have been nice to have something like the `CHROME_VERSION_EXTRA` variable to append a `-` + `version` to qutebrowser's config and data dir names.

**Comment by The-Compiler** [comment]
Another quick status update:

- `qt6-v2` has been rebased on the latest `master` now. There were quite some conflicts, so please let me know if you notice any regressions.
- I also managed to adjust build scripts to produce windows builds: [qutebrowser-2.5.0-amd64-qt6.exe](https://qutebrowser.org/tmp/qutebrowser-2.5.0-amd64-qt6.exe), [qutebrowser-2.5.0-windows-standalone-amd64-qt6.zip](https://qutebrowser.org/tmp/qutebrowser-2.5.0-windows-standalone-amd64-qt6.zip)
- No macOS builds yet. PyInstaller doesn't seem to work with Qt 6 on macOS yet. Will need to investigate...

---

@tinywrkb See #4102 re some ideas about having multiple named data profiles. Probably an argument (and/or envvar) would be a good idea to have, once that exists.

**Comment by The-Compiler** [comment]
This has a lot of important historic information which I'd rather not hide, but it's still a lot of stuff in this issue now. Thus, let's continue in #7202 where I tried to summarize everything that seemed still relevant.
